packages/devtools/src/utils.ts

```diff
@@ -30,3 +42,140 @@ export async function promiseMap<
+
+# Metadata access directives
+
+directive @documentAccount on FIELD_DEFINITION
+directive @documentVersion on FIELD_DEFINITION
+
+# Model definition
+
+enum ModelIndexType {
+  LIST # Account to multiple streams - default
+  SET # Account to multiple streams but only one reference
+  LINK # Account to single stream (IDX)
+  NONE # Indexing explicitly disabled
+}
+directive @model(
+  index: ModelIndexType = LIST
+  description: String
+  url: String
+) on OBJECT
+# When a model has index = SET, at least one field must have the @index directive
+directive @index on FIELD_DEFINITION
+
+# Account-based relations
+
+enum AccountLinkTarget {
+  SELF
+  OTHER
+  BOTH
+}
+directive @accountLink(
+  property: String!
+  target: AccountLinkTarget!
+) on FIELD_DEFINITION
+directive @accountReference on FIELD_DEFINITION
+directive @accountRelation(property: String!) on FIELD_DEFINITION
+
+# Model-based relations
+
+directive @modelReference(type: String!) on FIELD_DEFINITION
+directive @modelRelation(property: String!) on FIELD_DEFINITION
+`
+
+/** @internal */
+export function buildCompositeSchema(sdl: string): GraphQLSchema {
+  return (
+    lengthDirectiveTransformer(
+      modelDirectiveTransformer(
+        ipfsDirectiveTransformer(
+          makeExecutableSchema({typeDefs: [ceramicTypeDefs, sdl]})
+        )
+      )
+    )
+  ) 
+}
+
+function fieldSchemaFromFieldDefinition(
+  fieldDefinition: Record<string, any>
+  ): Record<string, any> | undefined {
+  const ceramicExtensions = fieldDefinition.extensions?.ceramicExtensions
+  if (ceramicExtensions?.length !== undefined) {
+    return {
+      type: 'string',
+      title: fieldDefinition.name,
+      maxLength: ceramicExtensions.length.max,
+      minLength: ceramicExtensions.length.min,
+    }
+  } else if (ceramicExtensions?.ipfs !== undefined) {
+    return {
+      type: 'ipfsUrlString',
+      title: fieldDefinition.name,
+      pattern: ceramicExtensions.ipfs.pattern,
+      maxLength: ceramicExtensions.ipfs.max,
+    }
+  } else {
+    return undefined
+  }
+}
+
+/** @internal */
+export function compositeDefinitionFromSchema(schema: string | GraphQLSchema): CompositeDefinition {
+  if (typeof schema === 'string') {
+    schema = buildCompositeSchema(schema)
+  }
+
+  let models: Record<string, ModelDefinition> = {}
+
+  mapSchema(schema, {
+    [MapperKind.OBJECT_TYPE]: (objectConfig: GraphQLObjectType) => {
+      const modelDirective = getDirective(schema as GraphQLSchema, objectConfig, 'model')?.[0];
+      if (modelDirective) {
+        const modelSchema:Record<string, any> = {}
+        for (const [fieldName, fieldDefinition] of Object.entries(objectConfig.getFields())) {
+          switch(fieldDefinition.type.toString()) {
+            case 'ceramicStringType!': // TODO: Is there a better way of introspecting the type?
+            case 'ceramicStringType':
+              modelSchema[fieldName] = fieldSchemaFromFieldDefinition(fieldDefinition)
```

Comment:
Seems strange to me need to apply this logic here, if I understand correctly this will only apply to models and not all objects?